src/wasm-interpreter.h

```diff
@@ -2228,6 +2228,107 @@ class ModuleInstance
     : ModuleInstanceBase(wasm, externalInterface) {}
 };
 
+static const Name NONSTANDALONE_FLOW("Binaryen|nonstandalone");
+
+typedef std::unordered_map<LocalGet*, Literals> GetValues;
+
+// Evaluates a standalone expression. Errors if we hit anything that can't be
+// evaluated.
+class StandaloneExpressionRunner
+  : public ExpressionRunner<StandaloneExpressionRunner> {
+  Module* module;
+
+  // map gets to constant values, if they are known to be constant
+  GetValues& getValues;
+
+  // Whether we are trying to precompute down to an expression (which we can do
+  // on say 5 + 6) or to a value (which we can't do on a local.tee that flows a
+  // 7 through it). When we want to replace the expression, we can only do so
+  // when it has no side effects. When we don't care about replacing the
+  // expression, we just want to know if it will contain a known constant.
+  bool replaceExpression;
```

Comment:
This seems semantically complex enough that having an enum to name the options would be helpful.